cloudinit/net/freebsd.py

```diff
@@ -0,0 +1,164 @@
+            for subnet in subnets:
+                if subnet.get('type') == 'static':
+                    gateway = subnet.get('gateway')
+                    if gateway:
+                        routes.append({
+                            'network': '0.0.0.0',
+                            'netmask': '0.0.0.0',
+                            'gateway': gateway})
+                    routes += subnet.get('routes', [])
+        route_cpt = 0
+        for route in routes:
+            network = route.get('network')
+            netmask = route.get('netmask')
+            gateway = route.get('gateway')
+            route_cmd = "-route %s/%s %s" % (network, netmask, gateway)
+            if not network:
+                continue
+            if network == '0.0.0.0':
+                rhel_util.update_sysconfig_file(
+                    util.target_path(target, self.rc_conf_fn), {
+                        'defaultrouter': gateway})
+            else:
+                rhel_util.update_sysconfig_file(
+                    util.target_path(target, self.rc_conf_fn), {
+                        'route_net%d' % route_cpt: route_cmd})
+                route_cpt += 1
+
+    def _write_resolve_conf(self, settings, target=None):
+        nameservers = settings.dns_nameservers
+        searchdomains = settings.dns_searchdomains
+        for interface in settings.iter_interfaces():
+            for subnet in interface.get("subnets", []):
+                if 'dns_nameservers' in subnet:
+                    nameservers.extend(subnet['dns_nameservers'])
+                if 'dns_search' in subnet:
+                    searchdomains.extend(subnet['dns_search'])
+        # Try to read the /etc/resolv.conf or just start from scratch if that
+        # fails.
+        try:
+            resolvconf = ResolvConf(util.load_file(self.resolv_conf_fn))
+            resolvconf.parse()
+        except IOError:
+            util.logexc(LOG, "Failed to parse %s, use new empty file",
+                        self.resolv_conf_fn)
+            resolvconf = ResolvConf('')
+            resolvconf.parse()
+
+        # Add some nameservers
+        for server in nameservers:
+            try:
+                resolvconf.add_nameserver(server)
+            except ValueError:
+                util.logexc(LOG, "Failed to add nameserver %s", server)
+
+        # And add any searchdomains.
+        for domain in searchdomains:
+            try:
+                resolvconf.add_search_domain(domain)
+            except ValueError:
+                util.logexc(LOG, "Failed to add search domain %s", domain)
+        util.write_file(
+            util.target_path(target, self.resolv_conf_fn),
+            str(resolvconf), 0o644)
+
+    def _write_network(self, settings, target=None):
+        self._write_ifconfig_entries(settings, target=None)
+        self._write_route_entries(settings, target=None)
+        self._write_resolve_conf(settings, target=None)
+
+        self.start_services()
```

Comment:
It would be nice to have a config toggle for this method.  Cloud-init has a sub-command 'cloud-init devel net-convert` which lets us attempt network conversion between formats; and in these cases, we don't want to actually restart services on the host;  much with the post-commands config.